sstables/sstable_set.cc

```diff
@@ -768,6 +768,393 @@ sstable_set_impl::create_single_key_sstable_reader(
     return make_combined_reader(schema, std::move(permit), std::move(readers), fwd, fwd_mr);
 }
 
+// Merges output from readers of TWCS sstables opened for a single partition query into a non-decreasing stream
+// of mutation fragments.
+//
+// Uses `min_position_reader_queue` to open the readers lazily as the read progresses through the partition.
+// A reader is opened only if we find that it may contain fragments for the currently inspected positions.
+//
+// Readers are closed as soon as we find that they were exhausted for the given partition query.
+//
+// Implements the `FragmentProducer` concept. However, `next_partition` and `fast_forward_to(partition_range)`
+// are not implemented and throw an error; the reader is only used for single partition queries.
+//
+// Assumes that the readers returned from `min_position_reader_queue` return fragments from the same partition.
+// Otherwise the behavior is unspecified.
+class time_series_reader_merger {
+    const schema_ptr _schema;
+    const reader_permit _permit;
+
+    // The key of the partition being queried.
+    const dht::decorated_key _key;
+
+    // Compares positions using *_schema.
+    const position_in_partition::less_compare _less;
+    const position_in_partition::equal_compare _eq;
+
+    // A queue of readers used to lazily open new readers as we progress through the partition.
+    // Before a batch with position `p` is returned, we ensure that all readers containing positions
+    // <= `p` are popped from the queue.
+    min_position_reader_queue _reader_queue;
+
+    // Owning container for the readers popped from _reader_queue.
+    // If we are sure that a reader is exhausted (all rows from the queried partition have been returned),
+    // we destroy and remove it from the container.
+    std::list<backed_reader> _all_readers;
+    using reader_iterator = std::list<backed_reader>::iterator;
+
+    // A reader together with the last fragment it returned.
+    // The iterator points to _all_readers.
+    // fragment.position() is <= than position of the next fragment returned by reader, if any.
+    struct peeked_reader {
```

Comment:
It would be nice if we used the same names we are already familiar with from the existing combined reader. The two implementations are very similar, they just use their respective selector at different places. Using the same name for stuff that is the same would help reduce the mental load.

Reply:
I didn't like the names used there, thought they are uninformative. The thing that had me banging my head against the wall for a day or two was `_next`, I couldn't wrap my head around its purpose (and to make it worse, among all other member variables, this one doesn't have a comment). But then I understood -- `_fragment_heap` has readers that we've recently peeked, and `_next` has readers for which we don't know the next fragment ("unpeeked") and we must peek them before doing anything else, since they might have earlier fragments.

Reply:
I don't like some of the names used there either, but kept them unchanged for the same reason. For other these names are just as perplexing :). We are not actually peaking the readers here, we are reading a new fragment from them.

Reply:
Imagine a reader that has a `peek()` method, which returns the next fragment but without popping it from the reader. E.g. like a vector's `back()` vs `pop_back()`. So you can peek the reader and decide if you want to remove the fragment now or not yet.

You can create such "peekable reader" by wrapping an existing reader and adding a new field, `optional<fragment> of`. The `peek` method would then check if `of` is filled, and if it is, return the fragment; otherwise, pull the fragment from the internal reader, put it in `of`, and return it. The popping method (`operator()`) would return the `of` fragment and clear `of` if it's filled, and if it's not, return it directly by popping the internal reader.

`peeked_reader` is just a very dumb implementation of the above-described "peekable reader". It serves the purpose of a reader for which we can check the next fragment that it will return, but without popping it: this is the `fragment` field of `peeked_reader`.

You are right we are not peeking the internal readers, we're pulling from them; but we're using the pulled fragment and the reader that we just pulled from to "create a new reader" (which I call `peeked_reader`) which we can then peek.

So maybe the name `peekable_reader` would be better.

Reply:
I don't have to imagine that -- there is a `flat_mutation_reader::peek()` and `flat_mutation_reader::peek_buffer()`, but you are not using them.

Reply:
BTW this just made me realize that `fragment_and_reader` from the old combined reader is obsolete, we could just use the above mentioned peek methods.

Reply:
Wait... what

Reply:
https://github.com/scylladb/scylla/blob/f893516e55cc9b230c00e7b094d0e4eeb77335b3/flat_mutation_reader.hh#L449-L465

Reply:
Removed `peeked_reader`, using `peek()` now